cluster-autoscaler/utils/drain/drain.go

```diff
@@ -106,102 +106,38 @@ func GetPodsForDeletionOnNodeDrain(
-				}
-			} else {
-				replicated = true
-			}
-		} else if pod_util.IsDaemonSetPod(pod) {
-			isDaemonSetPod = true
-			// don't have listener for other DaemonSet kind
-			// TODO: we should use a generic client for checking the reference.
-			if checkReferences && refKind == "DaemonSet" {
-				_, err := listers.DaemonSetLister().DaemonSets(controllerNamespace).Get(controllerRef.Name)
-				if apierrors.IsNotFound(err) {
-					return []*apiv1.Pod{}, []*apiv1.Pod{}, &BlockingPod{Pod: pod, Reason: ControllerNotFound}, fmt.Errorf("daemonset for %s/%s is not present, err: %v", pod.Namespace, pod.Name, err)
-				} else if err != nil {
-					return []*apiv1.Pod{}, []*apiv1.Pod{}, &BlockingPod{Pod: pod, Reason: UnexpectedError}, fmt.Errorf("error when trying to get daemonset for %s/%s , err: %v", pod.Namespace, pod.Name, err)
-				}
-			}
-		} else if refKind == "Job" {
-			if checkReferences {
-				job, err := listers.JobLister().Jobs(controllerNamespace).Get(controllerRef.Name)
-
-				// Assume the only reason for an error is because the Job is
-				// gone/missing, not for any other cause.  TODO(mml): something more
-				// sophisticated than this
-				if err == nil && job != nil {
-					replicated = true
-				} else {
-					return []*apiv1.Pod{}, []*apiv1.Pod{}, &BlockingPod{Pod: pod, Reason: ControllerNotFound}, fmt.Errorf("job for %s/%s is not available: err: %v", pod.Namespace, pod.Name, err)
-				}
-			} else {
-				replicated = true
-			}
-		} else if refKind == "ReplicaSet" {
-			if checkReferences {
-				rs, err := listers.ReplicaSetLister().ReplicaSets(controllerNamespace).Get(controllerRef.Name)
-
-				// Assume the only reason for an error is because the RS is
-				// gone/missing, not for any other cause.  TODO(mml): something more
-				// sophisticated than this
-				if err == nil && rs != nil {
-					if rs.Spec.Replicas != nil && *rs.Spec.Replicas < minReplica {
-						return []*apiv1.Pod{}, []*apiv1.Pod{}, &BlockingPod{Pod: pod, Reason: MinReplicasReached}, fmt.Errorf("replication controller for %s/%s has too few replicas spec: %d min: %d",
-							pod.Namespace, pod.Name, rs.Spec.Replicas, minReplica)
-					}
-					replicated = true
-				} else {
-					return []*apiv1.Pod{}, []*apiv1.Pod{}, &BlockingPod{Pod: pod, Reason: ControllerNotFound}, fmt.Errorf("replication controller for %s/%s is not available, err: %v", pod.Namespace, pod.Name, err)
-				}
-			} else {
+		if refKind != "" {
+			gv := strings.Split(controllerRef.APIVersion, "/")
+			// controllerRef doesn't have a namespace by design
+			// The controller/owner is either in the same namespace as the pod
+			// or it's assumed to be a cluster-scoped resource
+			// The assumption in the code below is that the controllerRef/owner of
+			// a pod resource will always be in the same namespace
+			// TODO: find a better way to handle this
+			l := listers.GenericListerFactory().GetLister(schema.GroupVersionResource{
```

Comment:
I am using generic lister for all resources even when we have a lister for some in-built resources like DaemonSets, StatefulSets etc., to simplify the code at the cost additional overhead of initializing on-demand listers (per namespace). Not sure if this is the best approach to do this. Would love to have any feedback on this :pray:

Reply:
How about making it an implementation detail of `GetLister`? If there's a dedicated lister already, it would use that, otherwise fall back to the dynamic one?

Reply:
I like the idea. Thank you for the suggestion!

Reply:
This idea might not work because there is no common `Lister` interface we can use as a return type. We could use `interface{}` as the return type but then the user would have to typecast the lister i.e., they'd have to know the lister type beforehand. 
e.g., 
 [`Lister` interface for dynamic/unstructured](https://github.com/kubernetes/autoscaler/blob/3955223eeeeea11089cc15779fcf1b4c5d8b54e9/cluster-autoscaler/vendor/k8s.io/client-go/dynamic/dynamiclister/interface.go#L24-L32) type is different from the [`ReplicaSet`'s Lister interface](https://github.com/kubernetes/autoscaler/blob/222efd9a4c4f5a933b79307d030c27d3c9913281/cluster-autoscaler/vendor/k8s.io/client-go/listers/apps/v1/replicaset.go#L28-L37) which is different from [Job's Lister interface](https://github.com/kubernetes/autoscaler/blob/222efd9a4c4f5a933b79307d030c27d3c9913281/cluster-autoscaler/vendor/k8s.io/client-go/listers/batch/v1/job.go#L28-L37) which is different from [StatefulSet's lister](https://github.com/kubernetes/autoscaler/blob/222efd9a4c4f5a933b79307d030c27d3c9913281/cluster-autoscaler/vendor/k8s.io/client-go/listers/apps/v1/statefulset.go#L28-L37) etc., 

I don't think there's a common Lister interface for in-built types but assuming there was, we would still have the same problem since unstructured's Lister interface is different.

Reply:
I wonder if there's an elegant way to handle this problem.

Reply:
Right, you'd have to wrap existing listers into the dynamic one and just use that return type.

Reply:
Let me think about this and get back. I am imagining something like a `BigLister` interface which would have the in-built listers along with the dynamic lister.

Reply:
> Right, you'd have to wrap existing listers into the dynamic one and just use that return type.

Not sure if I understand this correctly. 

I am guessing the idea is to change the return-type for built-in listers (deployments, statefulsets etc.,) with dynamic lister 

e.g., 
```go
// ListerRegistry is a registry providing various listers to list pods or nodes matching conditions
type ListerRegistry interface {
	AllNodeLister() NodeLister
	ReadyNodeLister() NodeLister
	ScheduledPodLister() PodLister
	UnschedulablePodLister() PodLister
	PodDisruptionBudgetLister() PodDisruptionBudgetLister
	DaemonSetLister() v1appslister.DaemonSetLister
	ReplicationControllerLister() v1lister.ReplicationControllerLister
	JobLister() v1batchlister.JobLister
	ReplicaSetLister() v1appslister.ReplicaSetLister
	StatefulSetLister() v1appslister.StatefulSetLister
	GetLister(gvr schema.GroupVersionKind, namespace string) dynamiclister.Lister
}
```
https://github.com/kubernetes/autoscaler/blob/13a3bf943d456cfdcdd266ffbae0a0231533b8c5/cluster-autoscaler/utils/kubernetes/listers.go#L49-L61

to 
```go
// ListerRegistry is a registry providing various listers to list pods or nodes matching conditions
type ListerRegistry interface {
	AllNodeLister() dynamiclister.Lister
	ReadyNodeLister() dynamiclister.Lister
	ScheduledPodLister() dynamiclister.Lister
	UnschedulablePodLister() dynamiclister.Lister
	PodDisruptionBudgetLister() dynamiclister.Lister
	DaemonSetLister() dynamiclister.Lister
	ReplicationControllerLister() dynamiclister.Lister
	JobLister() dynamiclister.Lister
	ReplicaSetLister() dynamiclister.Lister
	StatefulSetLister() dynamiclister.Lister
	GetLister(gvr schema.GroupVersionKind, namespace string) dynamiclister.Lister
}
```

Reply:
I like the idea.

One problem I see is, if we use `dynamic.Lister` as a return type (or wrapper type), accessing fields in the struct [won't be as straightforward](https://github.com/kubernetes/autoscaler/blob/13a3bf943d456cfdcdd266ffbae0a0231533b8c5/cluster-autoscaler/utils/kubernetes/listers.go#L267) as just doing something like `ob.Spec.replicas` ([which is possible for in-built types](https://github.com/kubernetes/autoscaler/blob/222efd9a4c4f5a933b79307d030c27d3c9913281/cluster-autoscaler/vendor/k8s.io/client-go/listers/apps/v1/statefulset.go#L33))

Reply:
I will think a bit more around https://github.com/kubernetes/autoscaler/pull/5419#discussion_r1099676412